datafusion/core/tests/sqllogictests/test_files/select.slt

```diff
@@ -214,3 +214,9 @@ select * from (select 1 a union all select 2) b order by a limit null;
 query I
 select * from (select 1 a union all select 2) b order by a limit 0;
 ----
+
+# select case when type coercion
+query I
+select CASE 10.5 WHEN 0 THEN 1 ELSE 10 END as col;
+----
+10
```

Comment:
Could you add some additional tests?

1. Alternate Syntax:
```
select CASE 
  WHEN 10 = 5 THEN 1 
  WHEN 'true' THEN 1 
  ELSE 10 
END as col;
```


Also negative cases like

```sql
select CASE 10.5 WHEN 'not a number' THEN 1 ELSE 10 END as col;
```

```sql
select CASE 
  WHEN 10 = 5 THEN 1 
  WHEN 'not boolean' THEN 1 
  ELSE 10 
END as col;
```

```sql
select CASE 
  WHEN 10 = 5 THEN 1 
  WHEN 5 THEN 'not a number'
  ELSE 10 
END as col;
```

```sql
select CASE 
  WHEN 10 = 5 THEN 1 
  WHEN 5 THEN 0
  ELSE 'goo' 
END as col;
```

Reply:
Is there a way in sqllogictests to change the config of DataFusion? Since for negative cases to produce error, would need to set `datafusion.optimizer.skip_failed_rules` to false, unless you mean for the negative case to fail as in the original issue?

Otherwise can add the negative case as a unit test in actual code

Reply:
Yes, you can use the `SET` command to change config values. For example https://github.com/apache/arrow-datafusion/blob/7b90474adf6c1afbb4a5988a5677834b07347321/datafusion/core/tests/sqllogictests/test_files/information_schema.slt#L30

Reply:
Testing with sqllogictests turned out to be kinda finicky, as this PR is meant to enhance the type coercion, but not actually testing the actual casting itself (since some of those negative cases would pass the type coercion but fail at execution time due to actual contents of the data). So I figured it'd be easier to add more tests as unit tests for proper testing.